Grant contributions in profile (#3316)

e18r opened the pull request:
##### Description
In order to be able to show the grant contribution/subscription status of a user in her profile, we need to create a few new tabs. These tabs will show the activity related to contributing to, subscribing to, and unsubscribing from grants. However, this activity is not recorded at the moment, so the first thing to do is to record it. 

1. Created a new activity type for one-time grant contributions
2. Created a new foreign key for the object Activity, so that it can be associated with Contribution objects
3. Recorded the three types of grant activity: one-off contributions, recurring subscriptions and cancellations
4. Modified the activity query in the dashboard view so that it also fetches subscription activities
5. Created a new template that formats grant activities

##### Checklist

- [X] Read and conforms to the [contributor guidelines](https://docs.gitcoin.co/mk_contributors/)
- [x] Read and conforms to the [style guidelines](https://docs.gitcoin.co/mk_styleguide/)
- [x] Linter status: 100% pass
- [ ] Added relevant tests
- [ ] Tests pass
- [x] Code coverage has not decreased
- [x] Changes don't break existing behavior
- [x] Commit message follows [commit guidelines](https://docs.gitcoin.co/mk_contributors/#step-4-commit)
- [x] [Code documentation](https://docs.gitcoin.co/mk_contributors/#docstrings)
- [ ] Completed: Code review by core team and any requested changes

##### Affected core subsystem(s)

dashboard, grants

##### Refers/Fixes

Fixes: https://github.com/gitcoinco/web/issues/3316

##### Testing and Sign-off

I created a few Grants with a custom token on Rinkeby and with another user I contributed to them, subscribed and cancelled subscription. Working as expected.

###### Contributor

- [x] Read and followed the [Contributor Guidelines](https://docs.gitcoin.co/mk_contributors/)
- [x] Tested all changes **locally**
- [x] Verified existing functionality
- [x] Ran `make pytest` and code coverage didn't decrease
###### Reviewer

- [ ] Affirm contributor guidelines have been followed and requested changes made
- [ ] CI tests and linting pass
- [ ] No conflicts (migrations, files, etc)
- [ ] Regression tested against staging or local deployment

###### Funder

- [ ] Validated requested changes were made to specification
- [ ] Bounty payout released to the contributor

owocki commented on app/dashboard/models.py:
```diff
@@ -1554,6 +1554,7 @@ class Activity(SuperModel):
         ('update_grant', 'Updated Grant'),
         ('killed_grant', 'Cancelled Grant'),
         ('new_grant_contribution', 'Contributed to Grant'),
+        ('new_grant_subscription', 'Subscribed to Grant'),
```
killed_grant_contribution?
new_grant?

e18r replied:
If we want to differentiate one-off contributions from recurring subscriptions, we need this new activity type. I assumed that's what you wanted when you said

> grant activity feed items for 'new subscription', 'new contribution', and 'cancel subscription' on gitcoin profile

owocki replied:
i dont necessarily want to differentiate those things.  i'm just pointing out that you haven't added activity types for killed grant contributions or for new grants.

e18r replied:
because they already exist

owocki replied:
derp derp

owocki commented:
looks like a good start! let me know when ready for review

e18r commented on app/grants/views.py:
```diff
@@ -37,7 +37,7 @@
 from app.utils import get_profile
 from cacheops import cached_view
-from dashboard.models import Profile
+from dashboard.models import Profile, Activity
```
```suggestion
from dashboard.models import Activity, Profile
```

e18r commented:
Ready for review!

owocki commented on app/dashboard/models.py:
```diff
@@ -1585,6 +1586,12 @@ class Activity(SuperModel):
         on_delete=models.CASCADE,
         blank=True, null=True
     )
+    subscription = models.ForeignKey(
```
i think you need to `./manage.py makemigrations` to make a migration for this

e18r replied:
Definitely. I didn't include it because I thought you guys would wanna do it yourselves or something. But I'm definitely gonna commit the migration then.

owocki commented on app/grants/views.py:
```diff
@@ -472,6 +479,12 @@ def subscription_cancel(request, grant_id, grant_slug, subscription_id):
         subscription.cancel_tx_id = request.POST.get('sub_cancel_tx_id', '')
         subscription.active = False
         subscription.save()
+        activity_kwargs = {
```
i see 
```
Activity.objects.create(**activity_kwargs)
```
s 

for `killed_grant_contribution`
and for `new_grant_contribution`

but i dont see anywhere activity object creation code for `killed_grant` or `new_grant`. 

are you going to add those?

e18r replied:
Only for `new_grant_subscription`, `new_grant_contribution` and `killed_grant_contribution` because these are the ones that are going to appear in the user's profile. The others are out of scope

owocki commented on app/grants/views.py:
```diff
@@ -387,13 +387,20 @@ def grant_fund(request, grant_id, grant_slug):
             subscription.contributor_profile = profile
             subscription.grant = grant
             subscription.save()
+            activity_kwargs = {
```
it'd be nice if the `activity_kwargs` and `Activity.objects.create` code was DRY'd.  maybe refactoring it into a `record_grant_activity_helper(activity_type, subscription)` would suffice

e18r replied:
OK

owocki commented on app/dashboard/migrations/0015_auto_20190308_1945.py:
```diff
@@ -0,0 +1,36 @@
+            name='subscription',
+            field=models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.CASCADE, related_name='activities', to='grants.Subscription'),
+        ),
+        migrations.AlterField(
```
i think you can remove the rest of the `AlterField`s from this migration since they have nothing to do with the scope of hte PR

e18r replied:
OK

owocki commented on app/grants/views.py:
```diff
@@ -604,3 +608,20 @@ def leaderboard(request):
     return TemplateResponse(request, 'grants/leaderboard.html', params)
+
+
+def record_grant_activity_helper(activity_type, subscription, profile):
```
this should also be called in grant.models.py:subscription.successful_contribution for the `new_grant` activity type.

e18r replied:
@owocki, the `new_grant` activity type is meant for when a user creates a grant, not when they subscribe to it. So it has nothing to do with the feature I implemented.

owocki commented:
just left a few small cleanup comments.. this is mostly looking good; excited to get it merged!

have you tested on the frontend yet?  does it look okay on the profile and on the /activity page?

e18r commented:
@owocki it looks OK on the profile page. I just tested the /activity page and it doesn't look OK. This is the first time you mention that page. It's not part of the original request or any of its comments or the associated tweet, nor did it come up in the current thread until your last comment two days ago. I would say the /activity page is out of scope for this bounty. What do you think?

owocki commented:
> I just tested the /activity page and it doesn't look OK. This is the first time you mention that page. It's not part of the original request or any of its comments or the associated tweet, nor did it come up in the current thread until your last comment two days ago. I would say the /activity page is out of scope for this bounty. What do you think?

Well, I can't ship it if it's going to cause a regression on other pages.  If we got technical I guess this would fall under the "Changes don't break existing behavior" checkbox.  But I dont think thats a super productive route to take the convo; I can also see why you'd want to manage your scope / time investment in this ticket.  I'd happily add another 0.3 ETH to the ticket if it means that `new_grant` activity items can be added, and that all of the added activities look good on the `/activity` page

e18r commented:
> But I dont think thats a super productive route to take the convo;

I agree!

> I can also see why you'd want to manage your scope / time investment in this ticket.

Thank you for understanding! This is also my way of learning how things work on Gitcoin.

>  I'd happily add another 0.3 ETH to the ticket if it means that `new_grant` activity items can be added, and that all of the added activities look good on the `/activity` page

Let's do it!

In order to prevent a similar situation in the future, do you also want `update_grant` and `killed_grant` to show in the profile page and in the activity page? How about an extra 0.2 ETH for these two?

owocki commented:
> In order to prevent a similar situation in the future, do you also want update_grant and killed_grant to show in the profile page and in the activity page? How about an extra 0.2 ETH for these two?

yes that works for me

e18r commented:
Done! Let me know if you have any comments.

Please don't forget to update the ticket to 1 ETH as agreed.

owocki commented:
updating the bounty now

e18r commented:
Thanks! Did you get the chance to check out my work? It's finished, in case you missed that.

owocki commented:
looking now

owocki commented:
two questions at first glance:
- merge conflicts need resolved
- how did you test this?

owocki commented:
didnt test, but code looks sane to me.  requesting review from the other devs now

e18r commented:
> two questions at first glance:
> 
>     * merge conflicts need resolved

I solved the conflict and then rebased... but something went wrong and now... I want to merge 332 commits into `gitcoinco:master`? help...

>     * how did you test this?

I created, modified, contributed to, subscribed to, unsubscribed from and deleted a few grants locally and checked that everything looked OK in the /profile and /activity pages.

e18r commented:
I did it! now everything went back to normal. Ouf...

owocki commented:
looking good!  i requested review from some other devs too

thelostone-mc commented on app/dashboard/templates/shared/profile_activities_grant.html:
```diff
@@ -0,0 +1,63 @@
+      <p>
+        <span>{{ activity.grant.network }}</span>
+      </p>
+    </div>
```
Could we indent this by one ?

e18r replied:
Sure!

thelostone-mc commented on app/dashboard/templates/shared/profile_activities_subscription.html:
```diff
@@ -0,0 +1,75 @@
+ {% if activity.subscription.grant.url %} href="{{ activity.subscription.grant.url }}"
+ {% elif activity.subscription.grant.reference_url %} href="{{ activity.subscription.grant.reference_url }}"
+ {% endif %}>
+  <div class="avatar-container col-1 justify-content-center special_tag hide_min_viewport">
```
Could we fix up the indent ?

thelostone-mc commented on app/retail/templates/shared/activity.html:
```diff
@@ -71,16 +103,21 @@
             <span>{{ row.metadata.value_in_eth }} {{row.token_name}}</span>
+            {% elif row.grant %}
+            {{ row.metadata.value_in_token | floatformat:2 }} / {{ row.metadata.amount_goal | floatformat:2 }}
+            {{ row.metadata.token_name }}
```
indent the above 2 lines by 1 ?

e18r replied:
Natürlich!

e18r replied:
I just discovered that I was using HTML mode in emacs to work on these templates. That's why I got these indentation flaws.

thelostone-mc commented on app/retail/templates/shared/activity.html:
```diff
@@ -93,7 +130,7 @@
       </a>
     {% else %}
-      <i class="far {{ row.icon }} last-icon"></i>
+      <i class="fa {{ row.icon }} last-icon"></i>
```
why was this change done ? 😅

e18r replied:
I just like it better like that, you know? ;) It's shorter, simpler, more beautiful.

thelostone-mc replied:
Could you undo this change ? `fa` is retired in `v5` of `font-awesome` which is what we are in now

e18r replied:
Sure. I'm sorry, I thought you were joking.

thelostone-mc replied:
Haha no worries ^_^

e18r commented:
@thelostone-mc thanks for your review, will look into it tomorrow.

e18r commented:
@thelostone-mc, here are the changes you requested. Please let me know if you find anything else that unsettles you ;)

e18r commented:
@thelostone-mc @owocki if there's anything else you'd like me to change or improve, please let me know

thelostone-mc commented:
@e18r could you resolve the conflicts?  otherwise LGTM 

cc @danlipert

e18r commented:
@thelostone-mc @danlipert solved conflicts

danlipert commented:
@e18r Hey Emilio - looks like the tests are failing, probably the same migration issue as before. Can you please fix and @ me here to review? Thanks!

e18r commented:
Hi @danlipert, here's the requested conflict resolution.

This review dynamic seems to have become a vicious circle, as this is the 5th conflict resolution I've had to make for this pull request. So please hurry up so we can get out of it! Thanks!

e18r commented:
Hello @danlipert, have you got the time to review my work yet?

I apologize for my last comment. It was rude. It won't happen again. Please let me know if you encounter another merge conflict and I will gladly solve it.

e18r commented:
Hey @owocki @thelostone-mc please don't forget to make the payment whenever you feel satisfied with my work

owocki commented:
happy to pay it out once its tested + merged + deployed.  right now it still looks like the tests are failing

e18r commented:
@owocki I'd be happy to test it, merge it and deploy it for you. I only need the right credentials.

owocki commented:
can you start by getting the tests to pass?

owocki commented:
you can run them on your local with `make test`

e18r commented:
@owocki the tests fail with and without my changes.

Without: https://pastebin.com/mHYzTrWC
With: https://pastebin.com/n9xkiHYm

owocki commented:
per your pastebin, it looks like npm is messed up on your container...

per Travis CI, the tests are failing because of

```
E           django.core.management.base.CommandError: Conflicting migrations detected; multiple leaf nodes in the migration graph: (0027_auto_20190401_1349, 0024_grant_activity in dashboard).
E           To fix them run 'python manage.py makemigrations --merge'

```

which _is_ an issue with this PR

e18r commented:
@owocki no problem, I'll get it fixed right away.

e18r commented:
@owocki I solved the conflicting migrations. Please let me know if anything else needs to be done.

e18r commented on app/dashboard/models.py:
```diff
@@ -1692,12 +1711,13 @@ def view_props(self):
-            else:
-                activity['urled_title'] = activity.get('title')
-        activity['humanized_activity_type'] = self.humanized_activity_type
+            if 'category' not in obj or obj['category'] == 'bounty': # backwards-compatible for category-lacking metadata
```
Hi @danlipert. I don't think so, since the body of this if statement applies only to activities of type bounty, as opposed to activities of type grant, whereas line 1720 should apply for both kinds of activities.

e18r replied:
@danlipert what's your opinion on this?

danlipert replied:
@e18r Makes sense, my bad! Thanks!

e18r commented:
@owocki @thelostone-mc I think we're ready for merging!
